src/v/datalake/values_parquet.cc

```diff
@@ -0,0 +1,147 @@
+        return serde::parquet::int64_value{.val = v.val};
+    }
+    parquet_conversion_outcome operator()(iceberg::timestamptz_value v) {
+        return serde::parquet::int64_value{.val = v.val};
+    }
+    parquet_conversion_outcome operator()(iceberg::string_value v) {
+        return serde::parquet::byte_array_value{.val = std::move(v.val)};
+    }
+    parquet_conversion_outcome operator()(iceberg::uuid_value v) {
+        iobuf buffer;
+        buffer.append(v.val.uuid().data, v.val.uuid().size());
+        return serde::parquet::fixed_byte_array_value{.val = std::move(buffer)};
+    }
+    parquet_conversion_outcome operator()(iceberg::fixed_value v) {
+        return serde::parquet::fixed_byte_array_value{.val = std::move(v.val)};
+    }
+    parquet_conversion_outcome operator()(iceberg::binary_value v) {
+        return serde::parquet::byte_array_value{.val = std::move(v.val)};
+    }
+    parquet_conversion_outcome operator()(iceberg::decimal_value v) {
+        iobuf buffer;
+        // big endian, starts from msb
+        auto low_part = ss::cpu_to_be(Int128Low64(v.val));
+        auto high_part = ss::cpu_to_be(Int128High64(v.val));
+        buffer.append(
+          reinterpret_cast<const char*>(&high_part), sizeof(int64_t));
+        buffer.append(
+          reinterpret_cast<const char*>(&low_part), sizeof(uint64_t));
+
+        return serde::parquet::fixed_byte_array_value{.val = std::move(buffer)};
+    }
+};
+struct value_converting_visitor {
+    ss::future<parquet_conversion_outcome>
+    operator()(iceberg::primitive_value value) {
+        co_return std::visit(
+          primitive_value_converting_visitor{}, std::move(value));
+    }
+    ss::future<parquet_conversion_outcome>
+    operator()(std::unique_ptr<iceberg::struct_value> value) {
+        serde::parquet::group_value group;
+        group.reserve(value->fields.size());
+        for (auto& field : value->fields) {
+            if (!field.has_value()) {
+                group.emplace_back(serde::parquet::null_value{});
+                continue;
+            }
+            auto result = co_await to_parquet_value(std::move(*field));
+            if (result.has_error()) {
+                co_return result.error();
+            }
+
+            group.emplace_back(std::move(result.value()));
+        }
+        co_return group;
+    }
+    ss::future<parquet_conversion_outcome>
+    operator()(std::unique_ptr<iceberg::list_value> list) {
+        serde::parquet::repeated_value repeated;
+        repeated.reserve(list->elements.size());
+        for (auto& element : list->elements) {
+            if (!element.has_value()) {
+                repeated.emplace_back(serde::parquet::null_value{});
+                continue;
+            }
+            auto result = co_await to_parquet_value(std::move(*element));
+            if (result.has_error()) {
+                co_return result.error();
+            }
+            repeated.emplace_back(std::move(result.value()));
```

Comment:
Element also needs to be wrapped in a struct because you can't nest repeated.

The schema for a list element looks like:

```
<list-repetition> group <name> (LIST) {
  repeated group list {
    <element-repetition> <element-type> element;
  }
}
```

So the thing you return is `name`. The thing that repeats is a `group` and the iceberg list element is `element`. So just wrap each element in `group` and that works.

Reply:
it is pretty confusing as `repeated_element` is already a struct, i thought that this is the group you mention

Reply:
Yeah it's a little confusing I agree. `repeated` is a flag on a node in the schema, not an actual type of node, where we can't really do that in C++. So basically what happens is all repeated fields are wrapped in `repeated_value`, but otherwise you still need all the normal levels of `group` nodes. The extra struct of repeated_element is mostly just to work around being able to forward declare the variant alias - because in parquet you have to be able to represent repeated primitive values, even if we don't use that when translating from iceberg.

Reply:
thanks for clarifying that, fixed